libnmstate/nm/connection.py

```diff
@@ -249,34 +258,77 @@ def _get_activation_metadata(self):
 
         return activation_type, activation_object
 
-    def waitfor_active_connection_async(self, ac):
-        ac.handlers.add(
-            ac.nm_active_connection.connect(
-                'state-changed', self._waitfor_active_connection_callback, ac
+    def wait_active_async(self):
+        GLib.timeout_add(
+            ACTIVATE_CHECK_INTERNAL,
+            ConnectionProfile._active_timeout_callbak,
+            self,
+        )
+
+    @staticmethod
+    def _active_timeout_callbak(self):
+        if self._activate_time > ACTIVATE_TIMEOUT:
+            self._mainloop.quit(
+                'Activation timeout({} seconds) on device {}, quitting'.format(
+                    ACTIVATE_TIMEOUT, self.devname
+                )
+            )
+
+        nm_act_con = get_device_active_connection(self.nmdevice)
+
+        if not nm_act_con:
```

Comment:
If the ac does not exist, what creates it?

Reply:
That's internal code of libnm, you may ask them.